docs/src/index.md

```diff
@@ -0,0 +1,177 @@
+
+```jldoctest
+julia> f = Fill("hello", 2,5)
+2×5 Fill{String}, with entries equal to hello
+
+julia> Array(f)
+2×5 Matrix{String}:
+ "hello"  "hello"  "hello"  "hello"  "hello"
+ "hello"  "hello"  "hello"  "hello"  "hello"
+```
+
+### Conversion to a sparse form
+
+These `Fill` array types may be converted to sparse arrays as well, which might be useful in certain cases
+```jldoctest sparse
+julia> using SparseArrays
+
+julia> z = Zeros{Int}(2,2)
+2×2 Zeros{Int64}
+
+julia> sparse(z)
+2×2 SparseMatrixCSC{Int64, Int64} with 0 stored entries:
+ ⋅  ⋅
+ ⋅  ⋅
+```
+Note, however, that most `Fill` arrays are not sparse, despite being lazily evaluated.
+
+These types have methods that perform many operations efficiently, including elementary algebra operations like multiplication and addition, as well as linear algebra methods like `norm`, `adjoint`, `transpose` and `vec`.
+
+### Custom axes
+
+The various `Fill` equivalents all support offset or custom axes, where instead of the size, one may pass a `Tuple` of axes. So, for example, one may use a `SOneTo` axis from [`StaticArrays.jl`](https://github.com/JuliaArrays/StaticArrays.jl) to construct a statically sized `Fill`.
+
+```jldoctest
+julia> using StaticArrays
+
+julia> f = Fill(2, (SOneTo(4), SOneTo(5)))
+4×5 Fill{Int64, 2, Tuple{SOneTo{4}, SOneTo{5}}} with indices SOneTo(4)×SOneTo(5), with entries equal to 2
+```
+
+The size of such an array would be known at compile time, permitting compiler optimizations.
+
+We may construct infinite fill arrays by passing infinite-sized axes, see [`InfiniteArrays.jl`](https://github.com/JuliaArrays/InfiniteArrays.jl).
+
+### Other lazy types
+
+A lazy representation of an identity matrix may be constructured using `Eye`. For example, a `4x4` identity matrix with `Float32` elements may be constructed as
+
+```jldoctest sparse
+julia> id = Eye{Float32}(4)
+4×4 Eye{Float32}
+
+julia> Array(id)
+4×4 Matrix{Float32}:
+ 1.0  0.0  0.0  0.0
+ 0.0  1.0  0.0  0.0
+ 0.0  0.0  1.0  0.0
+ 0.0  0.0  0.0  1.0
+
+julia> sparse(id)
+4×4 SparseMatrixCSC{Float32, Int64} with 4 stored entries:
+ 1.0   ⋅    ⋅    ⋅
+  ⋅   1.0   ⋅    ⋅
+  ⋅    ⋅   1.0   ⋅
+  ⋅    ⋅    ⋅   1.0
+```
+
+Note that an `Eye` actually returns a `Diagonal` matrix, where the diagonal is a `Ones` vector.
+
+## Warning about map and broadcasting
+
+Broadcasting operations, and `map` and `mapreduce`, are also done efficiently, by evaluating the function being applied only once:
+
+```jldoctest
+julia> map(sqrt, Fill(4, 2,5))  # one evaluation, not 10, to save time
+2×5 Fill{Float64}, with entries equal to 2.0
+
+julia> println.(Fill(pi, 10))
+π
+10-element Fill{Nothing}, with entries equal to nothing
+```
+
+Notice that this will only match the behaviour of a dense matrix from `fill` if the function is pure. And that this shortcut is taken before any other fused broadcast:
+
+```julia
```

Comment:
Set `Random.seed!` so we can jldoctest these?

Reply:
Updated